cartographer/mapping/internal/optimization/optimization_problem_2d.cc

```diff
@@ -301,6 +350,64 @@ void OptimizationProblem2D::Solve(
     }
   }
 
+  // Add fixed frame pose constraints.
+  std::map<int, CeresPose> C_fixed_frames;
+  for (auto node_it = node_data_.begin(); node_it != node_data_.end();) {
+    const int trajectory_id = node_it->id.trajectory_id;
+    const auto trajectory_end = node_data_.EndOfTrajectory(trajectory_id);
+    if (!fixed_frame_pose_data_.HasTrajectory(trajectory_id)) {
+      node_it = trajectory_end;
+      continue;
+    }
+
+    const TrajectoryData& trajectory_data = trajectory_data_.at(trajectory_id);
+    bool fixed_frame_pose_initialized = false;
+    for (; node_it != trajectory_end; ++node_it) {
+      const NodeId node_id = node_it->id;
+      const NodeSpec2D& node_data = node_it->data;
+
+      const std::unique_ptr<transform::Rigid3d> fixed_frame_pose =
+          Interpolate(fixed_frame_pose_data_, trajectory_id, node_data.time);
+      if (fixed_frame_pose == nullptr) {
+        continue;
+      }
+
+      const Constraint::Pose constraint_pose{
+          *fixed_frame_pose, options_.fixed_frame_pose_translation_weight(),
+          options_.fixed_frame_pose_rotation_weight()};
+
+      if (!fixed_frame_pose_initialized) {
+        transform::Rigid3d fixed_frame_pose_in_map;
+        if (trajectory_data.fixed_frame_origin_in_map.has_value()) {
+          fixed_frame_pose_in_map =
+              trajectory_data.fixed_frame_origin_in_map.value();
+        } else {
+          fixed_frame_pose_in_map =
+              transform::Embed3D(node_data.global_pose_2d) *
+              constraint_pose.zbar_ij.inverse();
+        }
+        C_fixed_frames.emplace(
+            std::piecewise_construct, std::forward_as_tuple(trajectory_id),
+            std::forward_as_tuple(
+                transform::Rigid3d(
+                    fixed_frame_pose_in_map.translation(),
+                    Eigen::AngleAxisd(
+                        transform::GetYaw(fixed_frame_pose_in_map.rotation()),
+                        Eigen::Vector3d::UnitZ())),
+                nullptr,
+                absl::make_unique<ceres::AutoDiffLocalParameterization<
+                    YawOnlyQuaternionPlus, 4, 1>>(),
+                &problem));
+        fixed_frame_pose_initialized = true;
+      }
+
+      problem.AddResidualBlock(CreateAutoDiffSpaCostFunction(constraint_pose),
+                               nullptr,
+                               C_fixed_frames.at(trajectory_id).translation(),
```

Comment:
This is not quite right I think: Translation is 3 values for x, y, z, but in 2D we expect 3 value for x, y, yaw. This does not necessarily lead to bad results, but means that the optimized value for yaw is not in the correct place in the output.